src/main/java/io/github/deansg/jeocodio/GeocodioClient.java

```diff
@@ -132,28 +158,62 @@ private String formatFieldsParam(List<String> fields) {
     }
 
     private <T> CompletableFuture<T> sendAsync(HttpRequest httpRequest, Class<T> clazz) {
-        return httpClient.sendAsync(httpRequest, HttpResponse.BodyHandlers.ofString())
-                .thenApply(resp -> {
-                    var body = resp.body();
-                    if (resp.statusCode() != 200) {
-                        throw new GeocodioStatusCodeException(resp.statusCode(), body);
-                    }
-                    return body;
-                })
+        return httpClient.sendAsync(httpRequest, HttpResponse.BodyHandlers.ofInputStream())
+                .thenApply(this::readResponse)
                 .thenApply(str -> this.gson.fromJson(str, clazz));
     }
 
+    private String readResponse(HttpResponse<InputStream> resp) {
+        try (var inputStream = getResponseInputStream(resp)) {
+            String json = new String(inputStream.readAllBytes(), StandardCharsets.UTF_8);
+            if (resp.statusCode() != 200) {
+                throw new GeocodioStatusCodeException(resp.statusCode(), json);
+            }
+            return json;
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private InputStream getResponseInputStream(HttpResponse<InputStream> resp) throws IOException {
+        var inputStream = resp.body();
+        if (resp.headers().firstValue("Content-Encoding").orElse("").equals("gzip")) {
+            try {
+                inputStream = new GZIPInputStream(inputStream);
+            } catch (IOException e) {
+                inputStream.close();
+                throw e;
```

Comment:
With the outer try-with, I don't think this is necessary, but I suppose it doesn't hurt.

Reply:
I actually added this because of your first comment, so that in case of an exception during the creation itself we won't have a memory leak. If you don't think this try-catch is necessary I'm not really sure what was your concern in the first comment (GZip stream definitely will close the underlying stream in a "happy" flow)

Reply:
And actually, if there was an exception, the stream might not be open and the `inputStream.close()` might throw an exception. The more I think about it the more I would remove this explicit close and rely on the outer try-with-resources.

Reply:
Yeah, that way my bad. I didn't notice the outer try-with. I think this can be removed. All looks good from my end.